### PR TITLE
Improve settings menu display and remove theme menu

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
We talked about improving the settings menu and we mentioned that firefox pocket was a nice inspiration so I implemented it. The result looks like this:

Screenshot from 2022-05-11 23-59-53

You can test it [here](https://rustdoc.crud.net/imperio/settings-menu-display/doc/foo/index.html).

Only question I have is: should I re-assign the shortcut `T` to this setting menu now that the theme menu is gone? For now I simply removed it.

Important to be noted: the full settings page (at `settings.html`) is still rendered the same as currently.

r? @jsha

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Fixed eslint errors. :)

File: src/doc/rustdoc/src/how-to-read-rustdoc.md [hunk 1]
```diff
@@ -91,12 +84,16 @@ When typing in the search bar, you can prefix your search term with a type
 followed by a colon (such as `mod:`) to restrict the results to just that
 kind of item. (The available items are listed in the help popup.)
 
+### Changing displayed theme
+
+You can change the displayed theme by opening the settings menu and then
```
Comment (jsha) [inline src/doc/rustdoc/src/how-to-read-rustdoc.md:89]:
```suggestion
You can change the displayed theme by opening the settings menu (the gear icon in the upper right) and then
```

File: src/doc/rustdoc/src/write-documentation/what-to-include.md [hunk 2]
```diff
@@ -109,7 +109,7 @@ rustdoc --extend-css custom.css src/lib.rs
 
 A good example of using this feature to create a dark theme is documented [on
 this blog].  Just remember, dark theme is already included in the rustdoc output
-by clicking on the paintbrush.  Adding additional options to the themes are
+by clicking on the wheel.  Adding additional options to the themes are
```
Comment (jsha) [inline src/doc/rustdoc/src/write-documentation/what-to-include.md:112]:
```suggestion
by clicking on the gear icon in the upper right.  Adding additional options to the themes are
```

File: src/librustdoc/html/static/css/rustdoc.css [hunk 3]
```diff
@@ -1413,9 +1412,32 @@ pre.rust {
...
+	z-index: 1;
+	display: block;
+	margin-top: 7px;
+	border-radius: 3px;
+	border: 1px solid;
+}
+#settings-menu #settings .setting-line {
+	margin: 0.6em;
+}
+#settings-menu #settings::before {
```
Comment (jsha) [inline src/librustdoc/html/static/css/rustdoc.css:1430]: Please add a comment indicating that this draws the little arrow connecting the settings menu to the gear icon.

File: src/librustdoc/html/static/js/settings.js [hunk 4]
```diff
@@ -246,25 +268,36 @@
     } else {
         // We replace the existing "onclick" callback.
         const settingsButton = getSettingsButton();
+        const settingsMenu = document.getElementById("settings");
+        window.hideSettings = function() {
+            // switchDisplayedElement(null);
```
Comment (jsha) [inline src/librustdoc/html/static/js/settings.js:273]: Remove commented-out code.

File: src/librustdoc/html/static/js/settings.js [hunk 5]
```diff
@@ -206,38 +206,60 @@
...
         return el;
     }
 
     const settingsMenu = buildSettingsPage();
 
+    function displaySettings() {
+        settingsMenu.style.display = "";
+    }
+
+    function checkIfElemIsInParents(elem, parent) {
```
Comment (jsha) [inline src/librustdoc/html/static/js/settings.js:244]: Naming issue: functions that return a boolean should be named as the condition when that boolean is true, not a verb describing their action. For instance: `elemIsInParent`.

File: src/librustdoc/html/static/js/settings.js [hunk 6]
```diff
@@ -246,25 +268,36 @@
...
+                displaySettings();
             }
         };
-        window.hideSettings = function() {
-            switchDisplayedElement(null);
-        };
+        settingsButton.onblur = blurHandler;
+        settingsButton.querySelector("a").onblur = blurHandler;
+        onEachLazy(settingsMenu.querySelectorAll("input"), e => {
+            e.onblur = blurHandler;
```
Comment (jsha) [inline src/librustdoc/html/static/js/settings.js:290]: I think we usually use `el` to name elements (`e` is usually an Event).

**Review by jsha — CHANGES_REQUESTED** [review]
Looks great! The UI interaction works very nicely. All tab selection works as it should, plus opening/closing, clicking outside the dropdown to dismiss, hitting Escape, etc.

The CSS on `settings.html` is broken in your demo: https://rustdoc.crud.net/imperio/settings-menu-display/doc/settings.html.

Also once this lands we should consider revamping the Help interface to use the same styles. Right now it's weird that the two buttons bring up overlays with two different styles.

**Comment by GuillaumeGomez** [comment]
> The CSS on settings.html is broken in your demo: https://rustdoc.crud.net/imperio/settings-menu-display/doc/settings.html.

Indeed, https://github.com/rust-lang/rust/pull/96939 was merged after. I'll rebase on it so it's all clean. :)

File: src/librustdoc/html/static/js/settings.js (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/settings.js:290]: Good point!

File: src/librustdoc/html/static/js/settings.js (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/settings.js:244]: Good catch. This one survived longer than expected. :laughing:

**Comment by GuillaumeGomez** [comment]
Updated! Also updated the demo.

**Pull request merged**